bin/cli.rs

```diff
@@ -0,0 +1,126 @@
+#![allow(missing_docs)]
+use std::{
+    collections::BTreeMap,
+    fs::{create_dir_all, File},
+    io::BufReader,
+    num::NonZeroUsize,
+    thread,
+};
+
+use alloy_consensus::constants::KECCAK_EMPTY;
+use alloy_primitives::{Address, B256};
+use alloy_provider::{Provider, ProviderBuilder};
+use alloy_rpc_types::{BlockId, BlockTransactionsKind};
+use clap::{Parser, Subcommand};
+use pevm::{
+    chain::{PevmChain, PevmEthereum},
+    EvmAccount, EvmCode, Pevm, RpcStorage,
+};
+use reqwest::Url;
+use tokio::runtime;
+
+#[derive(Parser, Debug)]
+pub struct Cli {
+    #[clap(subcommand)]
+    pub subcmd: Commands,
+}
+
+#[derive(Subcommand, Debug)]
+pub enum Commands {
+    /// Snapshot a block to disk.
+    Snapshot {
+        /// The ID of the block.
+        block_id: BlockId,
+        /// The RPC url to connect to.
+        rpc_url: String,
+        /// The output directory to write the snapshot to.
+        output_dir: String,
+        /// Path to bytecodes.
+        bytecode_path: Option<String>,
+    },
+}
+
+pub fn main() {
+    let args = Cli::parse();
+
+    match args.subcmd {
+        Commands::Snapshot {
+            block_id,
+            rpc_url,
+            output_dir,
+            bytecode_path,
+        } => {
+            // Define provider.
+            let provider = ProviderBuilder::new().on_http(
+                Url::parse(&rpc_url)
+                    .unwrap_or_else(|_| panic!("Invalid RPC URL supplied: {rpc_url}")),
+            );
+
+            // Retrieve block to snapshot.
+            let runtime = runtime::Runtime::new().unwrap();
+            let block = runtime
+                .block_on(provider.get_block(block_id, BlockTransactionsKind::Full))
+                .unwrap_or_else(|e| panic!("Failed to fetch block from provider: {e}",))
+                .unwrap_or_else(|| panic!("No block found for ID: {:?}", block_id));
+
+            // TODO: We could support more specs here. This could be another CLI argument.
```

Comment:
I'm not too sure if it matters as I've only seen this in tests, but chainspec assumes that all hardforks are enabled, which is not a case for all EVM chains. Meaning that the pre-compiles pulled in based on the spec may not exist yet. This could lead to undefined behaviour at the VM level.

Reply:
Yes, let's leave a TODO to parameterize `chain` to later add support for OP, RISE, etc.